### ISSUE TITLE
"Algorithm negotiation fail" even though server and client have matching algorithms

### ISSUE THREAD

**Issue by tofi86** [mwiede/jsch#199]
Hi,

first of all: thanks for maintaining an up-to-date fork of `jsch`!

We encountered a problem where jsch fails with a `Algorithm negotiation fail` but when looking at the jsch debug output, there are matching algorithms between client and server:

```
INFO: Connecting to [REDACTED] port 22
INFO: Connection established
INFO: Remote version string: SSH-2.0-FileZillaProEnterpriseServer_1.5.2
INFO: Local version string: SSH-2.0-JSCH_0.2.3
INFO: CheckCiphers: chacha20-poly1305@openssh.com
INFO: CheckKexes: curve25519-sha256,curve25519-sha256@libssh.org,curve448-sha512
INFO: CheckSignatures: ssh-ed25519,ssh-ed448
INFO: ssh-ed25519 is not available.
INFO: ssh-ed448 is not available.
DEBUG: server_host_key proposal before removing unavailable algos is: ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
DEBUG: server_host_key proposal after removing unavailable algos is: ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
DEBUG: server_host_key proposal before known_host reordering is: ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
DEBUG: server_host_key proposal after known_host reordering is: ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
INFO: SSH_MSG_KEXINIT sent
INFO: SSH_MSG_KEXINIT received
INFO: kex: server: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha1,diffie-hellman-group1-sha1
INFO: kex: server: ssh-ed25519
INFO: kex: server: aes256-gcm@openssh.com,aes192-gcm@openssh.com,aes128-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc
INFO: kex: server: aes256-gcm@openssh.com,aes192-gcm@openssh.com,aes128-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc
INFO: kex: server: hmac-sha2-256,hmac-sha1-96,hmac-sha1
INFO: kex: server: hmac-sha2-256,hmac-sha1-96,hmac-sha1
INFO: kex: server: none
INFO: kex: server: none
INFO: kex: server: 
INFO: kex: server: 
INFO: kex: client: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256,ext-info-c
INFO: kex: client: ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
INFO: kex: client: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
INFO: kex: client: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
INFO: kex: client: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
INFO: kex: client: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
INFO: kex: client: none
INFO: kex: client: none
INFO: kex: client: 
INFO: kex: client: 
INFO: Disconnecting from [REDACTED] port 22
```

This is our (shortened) Java code for Java 11 and we have the latest jsch 0.2.3 version installed:

```java
JSch.setLogger(new JschLogger(log));
Jsch jsch = new JSch();
Session session = jsch.getSession("username", "host", 22);
session.setTimeout(10 * 1000);
Properties config = new Properties();
config.put("StrictHostKeyChecking", "no");
session.setConfig(config);
session.setPassword("password");
session.connect();
```

The `connect()` method fails with `com.jcraft.jsch.JSchException: Algorithm negotiation fail`.

I don't quite understand what the different `kex: server` and `kex: client` lines mean, but I read it as the algorithm negotiation part, beginning from the first lines of `kex: server` and `kex: client` meaning what algos the server and client offers.

However, when looking at the available and offered algorithms, we find similarities even in the first 3 algos `curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,...`.
But why can't Jsch then find a matching algorithm??

When trying to connect to the server from the commandline with `ssh -vv` this works and this is the debug log:

```
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: Applying options for *
debug2: resolve_canonicalize: hostname [REDACTED] is address
debug2: ssh_connect_direct
debug1: Connecting to [REDACTED][[REDACTED]] port 22.
debug1: Connection established.
debug1: identity file /home/[REDACTED]/.ssh/id_rsa type -1
debug1: identity file /home/[REDACTED]/.ssh/id_rsa-cert type -1
debug1: identity file /home/[REDACTED]/.ssh/id_dsa type -1
debug1: identity file /home/[REDACTED]/.ssh/id_dsa-cert type -1
debug1: identity file /home/[REDACTED]/.ssh/id_ecdsa type -1
debug1: identity file /home/[REDACTED]/.ssh/id_ecdsa-cert type -1
debug1: identity file /home/[REDACTED]/.ssh/id_ed25519 type -1
debug1: identity file /home/[REDACTED]/.ssh/id_ed25519-cert type -1
debug1: identity file /home/[REDACTED]/.ssh/id_xmss type -1
debug1: identity file /home/[REDACTED]/.ssh/id_xmss-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_7.9p1 Debian-10+deb10u2
debug1: Remote protocol version 2.0, remote software version FileZillaProEnterpriseServer_1.5.2
debug1: no match: FileZillaProEnterpriseServer_1.5.2
debug2: fd 3 setting O_NONBLOCK
debug1: Authenticating to [REDACTED]:22 as '[REDACTED]'
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug2: local client KEXINIT proposal
debug2: KEX algorithms: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256,diffie-hellman-group14-sha1
,ext-info-c
debug2: host key algorithms: ssh-ed25519-cert-v01@openssh.com,ssh-ed25519,ecdsa-sha2-nistp256-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,rsa-sha2-512-cert-v01@openssh.com,rsa-sha2-256-cert-v01@openssh.com,ssh-rsa-cert-v01@
openssh.com,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256,ssh-rsa
debug2: ciphers ctos: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
debug2: ciphers stoc: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
debug2: MACs ctos: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
debug2: MACs stoc: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
debug2: compression ctos: none,zlib@openssh.com,zlib
debug2: compression stoc: none,zlib@openssh.com,zlib
debug2: languages ctos:
debug2: languages stoc:
debug2: first_kex_follows 0
debug2: reserved 0
debug2: peer server KEXINIT proposal
debug2: KEX algorithms: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha1,diffie-hellman-group1-sha1
debug2: host key algorithms: ssh-ed25519
debug2: ciphers ctos: aes256-gcm@openssh.com,aes192-gcm@openssh.com,aes128-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc
debug2: ciphers stoc: aes256-gcm@openssh.com,aes192-gcm@openssh.com,aes128-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc
debug2: MACs ctos: hmac-sha2-256,hmac-sha1-96,hmac-sha1
debug2: MACs stoc: hmac-sha2-256,hmac-sha1-96,hmac-sha1
debug2: compression ctos: none
debug2: compression stoc: none
debug2: languages ctos:
debug2: languages stoc:
debug2: first_kex_follows 0
debug2: reserved 0
debug1: kex: algorithm: curve25519-sha256
debug1: kex: host key algorithm: ssh-ed25519
debug1: kex: server->client cipher: aes128-gcm@openssh.com MAC: <implicit> compression: none
debug1: kex: client->server cipher: aes128-gcm@openssh.com MAC: <implicit> compression: none
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug1: Server host key: ssh-ed25519 SHA256:6DS0aRJySHYxYsNRjNP6u37er1SUvPT87WAH18z1/b0
debug1: Host '[REDACTED]' is known and matches the ED25519 host key.
debug1: Found key in /home/[REDACTED]/.ssh/known_hosts:1
debug2: set_newkeys: mode 1
debug1: rekey after 4294967296 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug2: set_newkeys: mode 0
debug1: rekey after 4294967296 blocks
```

If I understand this correctly, the `ssh` connection uses `aes128-gcm@openssh.com` - an algorithm which is also available in the server and client list in Jsch.

Did we ran into a Jsch bug? Does someone here have another idea?

Thanks for having a look!

Best,
Tobias

**Comment by norrisjeremy** [comment]
Hi @tofi86,

The issue you are encountering is be because the server only supports the `ssh-ed25519` type host key algorithm, but JSch only supports this algorithm in Java 15+ or if you have added Bouncy Castle to your classpath:
```
INFO: CheckSignatures: ssh-ed25519,ssh-ed448
INFO: ssh-ed25519 is not available.
INFO: ssh-ed448 is not available.
...
INFO: kex: server: ssh-ed25519
...
INFO: kex: client: ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
```

Thanks,
Jeremy

**Comment by tofi86** [comment]
Hi @norrisjeremy,

thanks for the hint! So it's all about the Host Key Algorithm in this case, and the second kex: server/client lines are for the host key algos?

I just checked with the BouncyCastle lib ion my classpath, but the debug output is exactly the same:

_(shortened output from the IntelliJ console)_
```
/Users/REDACTED/.sdkman/candidates/java/11.0.13-tem/bin/java -ea -Dspring.profiles.active=prototype,mysql,hibernate,debug,surefire
-Djava.util.logging.config.file=src/test/resources/logging.properties -Duser.language=en
-Didea.test.cyclic.buffer.size=1048576 -Dfile.encoding=UTF-8
-javaagent:/Users/REDACTED/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/222.3739.54/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=50146:/Users/REDACTED/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/222.3739.54/IntelliJ IDEA.app/Contents/bin
-classpath ...:/Users/REDACTED/.m2/repository/org/bouncycastle/bcprov-jdk18on/1.71.1/bcprov-jdk18on-1.71.1.jar:/Users/REDACTED/.m2/repository/com/github/mwiede/jsch/0.2.3/jsch-0.2.3.jar:...
com.intellij.rt.junit.JUnitStarter -ideVersion5 -junit4 de.REDACTED.REDACTED.NetworkConnectionTest,testConnection

...
INFO: CheckSignatures: ssh-ed25519,ssh-ed448
INFO: ssh-ed25519 is not available.
INFO: ssh-ed448 is not available.
...
INFO: kex: server: ssh-ed25519
...
INFO: kex: client: ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
```

Do I have to configure BouncyCastle anywhere else for Jsch?

Thanks, Tobias

**Comment by norrisjeremy** [comment]
Hi @tofi86,

There are multiple parts of the algorithm negotiation (see [RFC4253 7.1](https://datatracker.ietf.org/doc/html/rfc4253#section-7.1) for more details).

What JSch is logging is:
```
INFO: kex: server: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha1,diffie-hellman-group1-sha1
^^^^^^^^^ server advertised kex algorithms ^^^^^^^^
INFO: kex: server: ssh-ed25519
^^^^^^^^^ server advertised host key algorithms ^^^^^^^^
INFO: kex: server: aes256-gcm@openssh.com,aes192-gcm@openssh.com,aes128-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc
^^^^^^^^^ server advertised client-to-server cipher algorithms ^^^^^^^^
INFO: kex: server: aes256-gcm@openssh.com,aes192-gcm@openssh.com,aes128-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc
^^^^^^^^^ server advertised server-to-client cipher algorithms ^^^^^^^^
INFO: kex: server: hmac-sha2-256,hmac-sha1-96,hmac-sha1
^^^^^^^^^ server advertised client-to-server mac algorithms ^^^^^^^^
INFO: kex: server: hmac-sha2-256,hmac-sha1-96,hmac-sha1
^^^^^^^^^ server advertised server-to-client mac algorithms ^^^^^^^^
INFO: kex: server: none
^^^^^^^^^ server advertised client-to-server compression algorithms ^^^^^^^^
INFO: kex: server: none
^^^^^^^^^ server advertised server-to-client compression algorithms ^^^^^^^^
INFO: kex: server: 
^^^^^^^^^ server advertised client-to-server languages ^^^^^^^^
INFO: kex: server: 
^^^^^^^^^ server advertised server-to-client languages ^^^^^^^^
INFO: kex: client: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256,ext-info-c
^^^^^^^^^ client (JSch) advertised kex algorithms ^^^^^^^^
INFO: kex: client: ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
^^^^^^^^^ client (JSch) advertised host key algorithms ^^^^^^^^
INFO: kex: client: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
^^^^^^^^^ client (JSch) advertised client-to-server cipher algorithms ^^^^^^^^
INFO: kex: client: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
^^^^^^^^^ client (JSch) advertised server-to-client cipher algorithms ^^^^^^^^
INFO: kex: client: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
^^^^^^^^^ client (JSch) advertised client-to-server mac algorithms ^^^^^^^^
INFO: kex: client: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
^^^^^^^^^ client (JSch) advertised server-to-client mac algorithms ^^^^^^^^
INFO: kex: client: none
^^^^^^^^^ client (JSch) advertised client-to-server compression algorithms ^^^^^^^^
INFO: kex: client: none
^^^^^^^^^ client (JSch) advertised server-to-client compression algorithms ^^^^^^^^
INFO: kex: client: 
^^^^^^^^^ client (JSch) advertised client-to-server languages ^^^^^^^^
INFO: kex: client: 
^^^^^^^^^ client (JSch) advertised server-to-client languages ^^^^^^^^
```

Each of these algorithm lists are compared for one or more shared (matching) algorithms between the server and client (JSch).
If any of them do not contain at least one shared (matching) algorithm, the SSH connection will fail with `Algorithm negotiation fail`.

I'm puzzled why your application isn't working with Bouncy Castle added to the class path with your IntelliJ, as I believe it should be based the info you provided.

Are you able to construct some sort of standalone test that doesn't run under the auspices of IntelliJ to see if it works for you or not?

Thanks,
Jeremy

**Comment by tofi86** [comment]
@norrisjeremy thanks for the detailed explanation of the Jsch logging! I really appreciate that! 👍 

I will try to create a smaller sample program outside of our application tomorrow to check outside of IntelliJ and the Unit test environment. I may be able to share the code, but unfortunately I cannot share nor mock the target server environment.

@theit While doing further research about Bouncy Castle I also stumbled upon `Security.addProvider()` and tried out exactly your code fragment but with no luck. However, I did place it right before creating the Jsch instance which is maybe not early enough in the application lifecycle. But on the other hand, my unit test doesn't run the whole application server, and only the NetworkConnection class which is a pretty simple PoJo so I guess it should have had an impact if it was used at all.

**Comment by norrisjeremy** [comment]
Hi @tofi86,

I don't believe any calls to `Security.addProvider()` should be necessary.

I'm curious to understand a standalone case outside of IntelliJ because we have an integration test that runs fine ([AlgorithmsIT#testBCEd25519()](https://github.com/mwiede/jsch/blob/master/src/test/java/com/jcraft/jsch/AlgorithmsIT.java#L216))
I believe all Maven does is add Bouncy Castle the class path since it's a declared dependency in the [pom.xml](https://github.com/mwiede/jsch/blob/master/pom.xml#L82).

Could you perhaps try to manually coerce JSch to use Bouncy Castle for `ssh-ed25519`, by adding something like this explicitly to your application?
```
session.setConfig("keypairgen.eddsa", "com.jcraft.jsch.bc.KeyPairGenEdDSA");
session.setConfig("ssh-ed25519", "com.jcraft.jsch.bc.SignatureEd25519");
```

Thanks,
Jeremy

**Comment by tofi86** [comment]
Just adding the Bouncy Castle dependency to the classpath works. That's all which needs to be done.

For solving this, I created a small standalone project which worked fine outside our application. So I spent the morning debugging the dependencies in this huge legacy project (with ~ 40 maven modules and inter-dependencies and such) and found out that some former genius colleague manually added parts of the Bouncy Castle project in a a 'org.bouncycastle' source package in one of the maven modules 9 years ago - instead of adding it as a maven dependency 🙄  This source package seemed to have higher precendence on the classpath over the maven dependency – and it didn't have the algorithms we expected as it was only imported partially.

After refactoring this mess it works now...

Sigh... 😩

However, thanks for your help here! 👏
